File: rs/moq/src/ietf/subscribe.rs
//! IETF moq-transport-07 subscribe messages

use std::borrow::Cow;

use crate::{coding::*, Path};

// We only support Latest Group (0x1)
const FILTER_TYPE: u8 = 0x01;

// We only support Group Order descending (0x02)
const GROUP_ORDER: u8 = 0x02;

/// Subscribe message (0x03)
/// Sent by the subscriber to request all future objects for the given track.
#[derive(Clone, Debug)]
pub struct Subscribe<'a> {
	pub subscribe_id: u64,
	pub track_alias: u64,
	pub track_namespace: Path<'a>,
	pub track_name: Cow<'a, str>,
	pub subscriber_priority: u8,
}

impl<'a> Message for Subscribe<'a> {
	fn decode<R: bytes::Buf>(r: &mut R) -> Result<Self, DecodeError> {
		let subscribe_id = u64::decode(r)?;
		let track_alias = u64::decode(r)?;

		// Decode namespace (tuple of strings)
		let track_namespace = decode_namespace(r)?;

		let track_name = Cow::<str>::decode(r)?;
		let subscriber_priority = u8::decode(r)?;

		let group_order = u8::decode(r)?;
		if group_order != 0 && group_order != GROUP_ORDER {
			return Err(DecodeError::InvalidValue);
		}

		let filter_type = u8::decode(r)?;
		if filter_type != FILTER_TYPE {
			return Err(DecodeError::InvalidValue);
		}

		let num_params = u8::decode(r)?;
		if num_params != 0 {
			return Err(DecodeError::InvalidValue);
		}

		Ok(Self {
			subscribe_id,
			track_alias,
			track_namespace,
			track_name,
			subscriber_priority,
		})
	}

	fn encode<W: bytes::BufMut>(&self, w: &mut W) {
		self.subscribe_id.encode(w);
		self.track_alias.encode(w);
		encode_namespace(w, &self.track_namespace);
		self.track_name.encode(w);
		self.subscriber_priority.encode(w);
		GROUP_ORDER.encode(w);
		FILTER_TYPE.encode(w);
		0u8.encode(w); // no parameters
	}
}

/// SubscribeOk message (0x04)
#[derive(Clone, Debug)]
pub struct SubscribeOk {
	pub subscribe_id: u64,
	/// Largest group/object ID tuple
	pub largest: Option<(u64, u64)>,
}

impl Message for SubscribeOk {
	fn encode<W: bytes::BufMut>(&self, w: &mut W) {
		self.subscribe_id.encode(w);
		0u8.encode(w); // expires = 0
		GROUP_ORDER.encode(w);

		if let Some((group, object)) = self.largest {
			1u8.encode(w); // content exists
			group.encode(w);
			object.encode(w);
		} else {
			0u8.encode(w); // no content
		}

		0u8.encode(w); // no parameters
	}

	fn decode<R: bytes::Buf>(r: &mut R) -> Result<Self, DecodeError> {
		let subscribe_id = u64::decode(r)?;

		let expires = u64::decode(r)?;
		if expires != 0 {
			return Err(DecodeError::InvalidValue);
		}

		let _group_order = u8::decode(r)?; // Don't care about group order

		let mut largest = None;
		let content_exists = u8::decode(r)?;
		if content_exists == 1 {
			let group = u64::decode(r)?;
			let object = u64::decode(r)?;
			largest = Some((group, object));
		}

		let num_params = u8::decode(r)?;
		if num_params != 0 {
			return Err(DecodeError::InvalidValue);
		}

		Ok(Self { subscribe_id, largest })
	}
}

/// SubscribeError message (0x05)
#[derive(Clone, Debug)]
pub struct SubscribeError<'a> {
	pub subscribe_id: u64,
	pub error_code: u64,
	pub reason_phrase: Cow<'a, str>,
	pub track_alias: u64,
}

impl<'a> Message for SubscribeError<'a> {
	fn encode<W: bytes::BufMut>(&self, w: &mut W) {
		self.subscribe_id.encode(w);
		self.error_code.encode(w);
		self.reason_phrase.encode(w);
		self.track_alias.encode(w);
	}

	fn decode<R: bytes::Buf>(r: &mut R) -> Result<Self, DecodeError> {
		let subscribe_id = u64::decode(r)?;
		let error_code = u64::decode(r)?;
		let reason_phrase = Cow::<str>::decode(r)?;
		let track_alias = u64::decode(r)?;

		Ok(Self {
			subscribe_id,
			error_code,
			reason_phrase,
			track_alias,
		})
	}
}

/// Unsubscribe message (0x0a)
#[derive(Clone, Debug)]
pub struct Unsubscribe {
	pub subscribe_id: u64,
}

impl Message for Unsubscribe {
	fn encode<W: bytes::BufMut>(&self, w: &mut W) {
		self.subscribe_id.encode(w);
	}

	fn decode<R: bytes::Buf>(r: &mut R) -> Result<Self, DecodeError> {
		let subscribe_id = u64::decode(r)?;
		Ok(Self { subscribe_id })
	}
}

/// SubscribeDone message (0x0b)
#[derive(Clone, Debug)]
pub struct SubscribeDone<'a> {
	pub subscribe_id: u64,
	pub status_code: u64,
	pub reason_phrase: Cow<'a, str>,
	pub final_group_object: Option<(u64, u64)>,
}

impl<'a> Message for SubscribeDone<'a> {
	fn encode<W: bytes::BufMut>(&self, w: &mut W) {
		self.subscribe_id.encode(w);
		self.status_code.encode(w);
		self.reason_phrase.encode(w);

		if let Some((group, object)) = self.final_group_object {
			1u8.encode(w); // content exists
			group.encode(w);
			object.encode(w);
		} else {
			0u8.encode(w); // no content
		}
	}

	fn decode<R: bytes::Buf>(r: &mut R) -> Result<Self, DecodeError> {
		let subscribe_id = u64::decode(r)?;
		let status_code = u64::decode(r)?;
		let reason_phrase = Cow::<str>::decode(r)?;

		let mut final_group_object = None;
		let content_exists = u64::decode(r)?;
		if content_exists == 1 {
			let group = u64::decode(r)?;
			let object = u64::decode(r)?;
			final_group_object = Some((group, object));
		}

		Ok(Self {
			subscribe_id,
			status_code,
			reason_phrase,
			final_group_object,
		})
	}
}

/// Helper function to encode namespace as tuple of strings
fn encode_namespace<W: bytes::BufMut>(w: &mut W, namespace: &Path) {
	// Split the path by '/' to get individual parts
	let path_str = namespace.as_str();
	if path_str.is_empty() {
		0u64.encode(w);
	} else {
		let parts: Vec<&str> = path_str.split('/').collect();
		(parts.len() as u64).encode(w);
		for part in parts {
			part.encode(w);
		}
	}
}

/// Helper function to decode namespace from tuple of strings
fn decode_namespace<R: bytes::Buf>(r: &mut R) -> Result<Path<'static>, DecodeError> {
	let count = u64::decode(r)? as usize;

	if count == 0 {
		return Ok(Path::from(String::new()));
	}

	let mut parts = Vec::with_capacity(count.min(16));
	for _ in 0..count {
		let part = String::decode(r)?;
		parts.push(part);
	}

	Ok(Path::from(parts.join("/")))
}

File: js/moq/src/ietf/subscribe.ts
import type * as Path from "../path.ts";
import type { Reader, Writer } from "../stream.ts";
import * as Message from "./message.ts";
import * as Namespace from "./namespace.ts";
import { Parameters } from "./parameters.ts";

// we only support Group Order descending
const GROUP_ORDER = 0x02;

export class Subscribe {
	static id = 0x03;

	requestId: bigint;
	trackNamespace: Path.Valid;
	trackName: string;
	subscriberPriority: number;

	constructor(requestId: bigint, trackNamespace: Path.Valid, trackName: string, subscriberPriority: number) {
		this.requestId = requestId;
		this.trackNamespace = trackNamespace;
		this.trackName = trackName;
		this.subscriberPriority = subscriberPriority;
	}

	async #encode(w: Writer): Promise<void> {
		await w.u62(this.requestId);
		await Namespace.encode(w, this.trackNamespace);
		await w.string(this.trackName);
		await w.u8(this.subscriberPriority);
		await w.u8(GROUP_ORDER);
		await w.u8(1); // forward = true
		await w.u8(0x2); // LargestObject
		await w.u8(0); // no parameters
	}

	async encode(w: Writer): Promise<void> {
		return Message.encode(w, this.#encode.bind(this));
	}

	static async decode(r: Reader): Promise<Subscribe> {
		return Message.decode(r, Subscribe.#decode);
	}

	static async #decode(r: Reader): Promise<Subscribe> {
		const requestId = await r.u62();
		const trackNamespace = await Namespace.decode(r);
		const trackName = await r.string();
		const subscriberPriority = await r.u8();

		const groupOrder = await r.u8();
		if (groupOrder > 2) {
			throw new Error(`unknown group order: ${groupOrder}`);
		}

		const forward = await r.u8();
		if (forward !== 1) {
			throw new Error(`unsupported forward value: ${forward}`);
		}

		const filterType = await r.u8();
		if (filterType !== 0x1 && filterType !== 0x2) {
			throw new Error(`unsupported filter type: ${filterType}`);
		}

		await Parameters.decode(r); // ignore parameters

		return new Subscribe(requestId, trackNamespace, trackName, subscriberPriority);
	}
}

export class SubscribeOk {
	static id = 0x04;

	requestId: bigint;
	trackAlias: bigint;

	constructor(requestId: bigint, trackAlias: bigint) {
		this.requestId = requestId;
		this.trackAlias = trackAlias;
	}

	async #encode(w: Writer): Promise<void> {
		await w.u62(this.requestId);
		await w.u62(this.trackAlias);
		await w.u62(0n); // expires = 0
		await w.u8(GROUP_ORDER);
		await w.u8(0); // no largest group/object
		await w.u8(0); // no parameters
	}

	async encode(w: Writer): Promise<void> {
		return Message.encode(w, this.#encode.bind(this));
	}

	static async decode(r: Reader): Promise<SubscribeOk> {
		return Message.decode(r, SubscribeOk.#decode);
	}

	static async #decode(r: Reader): Promise<SubscribeOk> {
		const requestId = await r.u62();
		const trackAlias = await r.u62();
		const expires = await r.u62();
		if (expires !== BigInt(0)) {
			throw new Error(`unsupported expires: ${expires}`);
		}

		await r.u8(); // Don't care about group order

		const contentExists = await r.u8();
		if (contentExists === 1) {
			// Ignore largest group/object
			await r.u62();
			await r.u62();
		}

		await Parameters.decode(r); // ignore parameters

		return new SubscribeOk(requestId, trackAlias);
	}
}

export class SubscribeError {
	static id = 0x05;

	requestId: bigint;
	errorCode: number;
	reasonPhrase: string;

	constructor(requestId: bigint, errorCode: number, reasonPhrase: string) {
		this.requestId = requestId;
		this.errorCode = errorCode;
		this.reasonPhrase = reasonPhrase;
	}

	async #encode(w: Writer): Promise<void> {
		await w.u62(this.requestId);
		await w.u62(BigInt(this.errorCode));
		await w.string(this.reasonPhrase);
	}

	async encode(w: Writer): Promise<void> {
		return Message.encode(w, this.#encode.bind(this));
	}

	static async decode(r: Reader): Promise<SubscribeError> {
		return Message.decode(r, SubscribeError.#decode);
	}

	static async #decode(r: Reader): Promise<SubscribeError> {
		const requestId = await r.u62();
		const errorCode = Number(await r.u62());
		const reasonPhrase = await r.string();

		return new SubscribeError(requestId, errorCode, reasonPhrase);
	}
}

export class Unsubscribe {
	static readonly id = 0x0a;

	requestId: bigint;

	constructor(requestId: bigint) {
		this.requestId = requestId;
	}

	async #encode(w: Writer): Promise<void> {
		await w.u62(this.requestId);
	}

	async encode(w: Writer): Promise<void> {
		return Message.encode(w, this.#encode.bind(this));
	}

	static async decode(r: Reader): Promise<Unsubscribe> {
		return Message.decode(r, Unsubscribe.#decode);
	}

	static async #decode(r: Reader): Promise<Unsubscribe> {
		const requestId = await r.u62();
		return new Unsubscribe(requestId);
	}
}

// In draft-14, this message is renamed from SUBSCRIBE_DONE to PUBLISH_DONE
export class PublishDone {
	static readonly id = 0x0b;

	requestId: bigint;
	statusCode: number;
	reasonPhrase: string;

	constructor(requestId: bigint, statusCode: number, reasonPhrase: string) {
		this.requestId = requestId;
		this.statusCode = statusCode;
		this.reasonPhrase = reasonPhrase;
	}

	async #encode(w: Writer): Promise<void> {
		await w.u62(this.requestId);
		await w.u62(BigInt(this.statusCode));
		await w.string(this.reasonPhrase);
		await w.u62(BigInt(0)); // stream_count = 0 (unsupported)
	}

	async encode(w: Writer): Promise<void> {
		return Message.encode(w, this.#encode.bind(this));
	}

	static async decode(r: Reader): Promise<PublishDone> {
		return Message.decode(r, PublishDone.#decode);
	}

	static async #decode(r: Reader): Promise<PublishDone> {
		const requestId = await r.u62();
		const statusCode = Number(await r.u62());
		const reasonPhrase = await r.string();
		await r.u62(); // ignore stream_count

		return new PublishDone(requestId, statusCode, reasonPhrase);
	}
}

// Backward compatibility alias
export const SubscribeDone = PublishDone;
